api/grpc/chunk_merge.go

```diff
@@ -0,0 +1,160 @@
+//     https://opensource.org/licenses/Apache-2.0
+
+package grpc
+
+import (
+	"container/heap"
+	"encoding/binary"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/tsdb/chunks"
+	"github.com/thanos-io/thanos/pkg/store/labelpb"
+	"github.com/thanos-io/thanos/pkg/store/storepb"
+
+	"github.com/thanos-io/thanos-parquet-gateway/internal/encoding"
+	"github.com/thanos-io/thanos-parquet-gateway/search"
+)
+
+// ChunkResponse represents a series with its raw chunk data from a single shard
+type ChunkResponse struct {
+	Series     search.SeriesChunks
+	ShardIndex int
+}
+
+// LoserTreeChunkMerger efficiently merges chunk responses from multiple shards
+// using a loser tree (min-heap) for optimal performance
+type LoserTreeChunkMerger struct {
+	responses []ChunkResponse
+	heap      chunkResponseHeap
+}
+
+// NewLoserTreeChunkMerger creates a new merger for chunk responses
+func NewLoserTreeChunkMerger(responses []ChunkResponse) *LoserTreeChunkMerger {
+	merger := &LoserTreeChunkMerger{
+		responses: responses,
+		heap:      make(chunkResponseHeap, 0, len(responses)),
+	}
+
+	for i, resp := range responses {
+		if len(resp.Series.Chunks) > 0 {
+			merger.heap = append(merger.heap, &heapItem{
+				response: resp,
+				index:    i,
+			})
+		}
+	}
+	heap.Init(&merger.heap)
+
+	return merger
+}
+
+// MergeToStorePB converts merged chunk responses to storepb.Series format
+func (m *LoserTreeChunkMerger) MergeToStorePB() ([]*storepb.Series, error) {
+	var result []*storepb.Series
+
+	for len(m.heap) > 0 {
+		current := heap.Pop(&m.heap).(*heapItem)
+		currentLabels := current.response.Series.Lset
+
+		var sameLabelsResponses []*heapItem
+		sameLabelsResponses = append(sameLabelsResponses, current)
+
+		for len(m.heap) > 0 && labels.Equal(m.heap[0].response.Series.Lset, currentLabels) {
+			item := heap.Pop(&m.heap).(*heapItem)
+			sameLabelsResponses = append(sameLabelsResponses, item)
+		}
+
+		storeSeries, err := m.mergeChunksToStoreSeries(currentLabels, sameLabelsResponses)
+		if err != nil {
+			return nil, err
+		}
+
+		result = append(result, storeSeries)
+	}
+
+	return result, nil
+}
+
+// mergeChunksToStoreSeries combines chunks from multiple shards into a single storepb.Series
+func (m *LoserTreeChunkMerger) mergeChunksToStoreSeries(lset labels.Labels, items []*heapItem) (*storepb.Series, error) {
+	storeLabels := labelpb.ZLabelSet{Labels: zLabelsFromMetric(lset)}
+
+	var allChunks []chunks.Meta
+	for _, item := range items {
+		allChunks = append(allChunks, item.response.Series.Chunks...)
+	}
+
+	storeChunks, err := m.chunksToStorePB(allChunks)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storepb.Series{
+		Labels: storeLabels.Labels,
+		Chunks: storeChunks,
+	}, nil
+}
+
+// chunksToStorePB converts Prometheus chunks.Meta to storepb.AggrChunk format
+func (m *LoserTreeChunkMerger) chunksToStorePB(chunkMetas []chunks.Meta) ([]storepb.AggrChunk, error) {
```

Comment:
I'm not 100% sure, but reencoding should not be needed. Could we add an integration test against Thanos Select() implementation? Basically, run some kind of query against this server implementation.

Reply:
sure @GiedriusS make sense will  pass `chunkMeta.Chunk.Bytes()` directly to `storepb.Chunk.Data` without additional encoding